### ISSUE TITLE
XObjectExpression adapter causes stack overflow

### ISSUE THREAD

**Issue by poxenham-myp** [ClosedXML/ClosedXML#2482]
## Read and complete the full issue template

Do not randomly delete sections. They are here for a reason.

**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature
- [ ] Question

**Did you test against the latest CI build?**
- [x] Yes
- [ ] No

If you answered `No`, please test with [the latest development build](https://ci.appveyor.com/project/ClosedXML/ClosedXML/branch/develop/artifacts) first.

**Version of ClosedXML**

e.g. 0.104.1

**What is the current behavior?**
Reading the values from the range F95 to BA101 on the second tab of the attached file after it has been modified causes a stack overflow exception. See attached log.

**What is the expected behavior or new feature?**

Reading the values from the range F95 to BA101 on the second tab of the file attached works. This did work with v0.102.3.

NOTE: 
* Passing evaluateFormulae: true to SaveAs when saving the modified file prevents the stack overflow exception BUT the values in the F95 to BA101 area are NOT updated and the values are stale
* Not passing evaluateFormulae: true to SaveAs causes the exception, but opening the cleaned file DOES have the correct values in the F95 to BA101 area

**Is this a regression from the previous version?**

Yes. Works in v0.102.3

## Reproducibility
**This is an important section. Read it carefully. Failure to do so will cause a 'RTFM' comment.**

Without a code sample, it is unlikely that your issue will get attention. Don't be lazy. Do the effort and assist the developers to reproduce your problem. Code samples should be [minimal complete and verifiable](https://stackoverflow.com/help/mcve). Sample spreadsheets should be attached whenever applicable. Remove sensitive information.

**Code to reproduce problem:**
```c#
// See https://aka.ms/new-console-template for more information

using System.Diagnostics;
using ClosedXML.Excel;

static void CleanUpRocFile(XLWorkbook workbook)
{
    // Create a memory stream to hold the cleaned workbook
    MemoryStream cleanedStream = new MemoryStream();
    IXLWorksheet worksheet = workbook.Worksheet("Step 2.Roster of Care (ROC)");

    // Clean "Vacancy" & "Sleepover"
    for (int i = 21; i <= 90; i++)
    {
        // Cell at column E. colIndex = 5
        IXLCell vacancyCell = worksheet.Cell(i, 5);
        string vacancyValue = vacancyCell.GetValue<string>().ToLowerInvariant();
        string[] vacancyNeedToClean = { "vacancy", "vacant", "vacancies", "0" };

        // Cell at column BG. colIndex = 59
        IXLCell sleepOverCell = worksheet.Cell(i, "BG");
        var sleepOverValue = sleepOverCell.GetValue<string>().ToLowerInvariant();

        if (string.IsNullOrWhiteSpace(vacancyValue) || vacancyNeedToClean.Contains(vacancyValue))
        {
            // Clear cells from column F to BA. colIndex = 6 - 53
            worksheet.Range(worksheet.Cell(i, 6), worksheet.Cell(i, 53)).Clear(XLClearOptions.Contents);
        }
        else if (sleepOverValue == "sleepover")
        {
            // Clear cells from column F to Q. colIndex = 6 - 17
            worksheet.Range(worksheet.Cell(i, 6), worksheet.Cell(i, 17)).Clear(XLClearOptions.Contents);
            // Clear cells from column AX to BA. colIndex = 50 - 53
            worksheet.Range(worksheet.Cell(i, 50), worksheet.Cell(i, 53)).Clear(XLClearOptions.Contents);
        }
    }

    // Save the cleaned data to the MemoryStream
    workbook.SaveAs("CleanedExcelFile.xlsx");
    
}
XLWorkbook wb = new XLWorkbook("ROC.xlsx");
CleanUpRocFile(wb);

wb = new XLWorkbook("CleanedExcelFile.xlsx");

IXLWorksheet worksheet = wb.Worksheets.Worksheet("Step 2.Roster of Care (ROC)");

IXLRange rocDataRange = worksheet.Range(95, 6, 101, 53);


// foreach over enumerable

var cells = rocDataRange.Rows().Cells();
Stopwatch  stopwatch = new Stopwatch();
stopwatch.Start();
foreach (var cell in cells)
{
    Console.WriteLine(cell.GetDouble());
}

var result = rocDataRange.Rows().Select(row =>
{
    var values = row.Cells()
        .Select(cell => cell.IsEmpty() ? 0 : cell.GetDouble()).ToList();
    double fundedHours = Math.Round(values.Sum() / 2, 2, MidpointRounding.AwayFromZero);
    double minimumHours = Math.Round(values.Select(Math.Ceiling).Sum() / 2, 2, MidpointRounding.AwayFromZero);
    return (fund: fundedHours, min: minimumHours);
});

foreach (var (fund, min) in result)
{
    Console.WriteLine($"Funded: {fund}, Minimum: {min}");
}
stopwatch.Stop();
Console.WriteLine($"Elapsed time: {stopwatch.ElapsedMilliseconds} ms");
```

- [x] I attached a sample spreadsheet.  (You can drag files on to this issue)
[ROC.xlsx](https://github.com/user-attachments/files/17550531/ROC.xlsx)

**Comment by doug-w** [comment]
Following, I believe I have triggered this with a simple repro where one sheet refers to another in a VLOOKUP
IE: Sheet1!D2 contains:
=C2 + IFNA((VLOOKUP(B2, Sheet2!A2:D9, 2, FALSE) - 1) * 10, 0)
Calling
row.Cell(i).Value.ToString();
throws a stack overflow with XObjectExpression early in the stack.

If it would help I can see if I can shrink it down to a small repro.

**Comment by jahav** [comment]
I have decided to bite the bullet and scrap the whole adapter infrastructure for `Expression`-signature functions.

The only reason why it still exists is that I didn't want to reimplement/refactor ~190 functions that were written for previous `Expression` based engine. Switching engine was hard enough to do in reasonable time, plus it worked until I decided to switch from recursive evaluation to calc chain one (0.103). Changing the functions would take whole release (and will likely take basically all time for 0.105, [list of reimplementation PRs](https://github.com/ClosedXML/ClosedXML/pulls?q=is%3Apr+author%3Ajahav++milestone%3Av0.105.0+reimplement+)).

I could fix the problem in this issue, but Expression adapters inherently don't work because of previous CalcEngine design limitations and inconsistency (https://github.com/ClosedXML/ClosedXML/pull/1802#issuecomment-1150438496). Rather than trying to make legacy `Expression`-based functions work with `AnyValue` based engine, it is better to just replace it all (plus Expression-based functions are not always sufficiently faithful to how spec/Excel says they should behave, so it's an oportunity to make them more faithful).

**Comment by jahav** [comment]
Fixed through #2583 and all dependent PRs.
The provided sample now works (well, doesn't crash and provides same output as 0.102.3).

Also, it's, also it's faster

| 0.102.3 | 0.105.0 |
| ---------| ---------| 
|Funded: 52.65, Minimum: 64.5<br/>Funded: 51.4, Minimum: 65<br/>Funded: 58.25, Minimum: 72<br/>Funded: 48.65, Minimum: 60.5<br/>Funded: 58.25, Minimum: 72<br/>Funded: 54.1, Minimum: 67<br/>Funded: 59.3, Minimum: 72<br/>Elapsed time: **111 569 ms** |  Funded: 52.65, Minimum: 64.5<br/>Funded: 51.4, Minimum: 65<br/>Funded: 58.25, Minimum: 72<br/>Funded: 48.65, Minimum: 60.5<br/>Funded: 58.25, Minimum: 72<br/>Funded: 54.1, Minimum: 67<br/>Funded: 59.3, Minimum: 72<br/>Elapsed time: **4 933 ms** |
